src/main/java/KeyVaultManagedStorageSamples.java

```diff
@@ -34,72 +31,102 @@ public KeyVaultManagedStorageSamples() throws InterruptedException, ExecutionExc
-        Iterator<RoleDefinition> roleDefs = roleDefinitions.iterator();
-        RoleDefinition keyVaultRole = null;
-        while (roleDefs.hasNext()) {
-            RoleDefinition definition = roleDefs.next();
-            if (definition.roleName().equals("Storage Account Key Operator Service Role")) {
-                keyVaultRole = definition;
-                break;
-            }
+        Optional<RoleDefinition> operationDefinition =
+                azureResourceManager.accessManagement()
+                        .roleDefinitions().listByScope("/")
+                        .stream()
+                        .filter(roleDefinition -> "Storage Account Key Operator Service Role".equals(roleDefinition.roleName()))
+                        .findFirst();
+        if (operationDefinition.isPresent()) {
+            azureResourceManager.accessManagement().roleAssignments()
+                    .define(UUID.randomUUID().toString()) //Needs to be a UUID formatted String
+                    .forObjectId("93c27d83-f79b-4cb2-8dd4-4aa716542e74") //This is the Azure Key Vault Service Principal
+                    .withRoleDefinition(operationDefinition.get().id())
+                    .withScope(storageAccount.id())
+                    .create();
         }
 
-        azure.accessManagement().roleAssignments()
-                .define(UUID.randomUUID().toString()) //Needs to be a UUID formatted String
-                .forObjectId("93c27d83-f79b-4cb2-8dd4-4aa716542e74") //This is the Azure Key Vault Service Principal
-                .withRoleDefinition(keyVaultRole.id())
-                .withScope(storageAccount.id())
-                .create();
-
-        // Setting the storage acocunt can only be called by a user account with access to the keys of the storage account.
-        // Therefore, we grant the user that created the storage account access to the vault
-        // as well as the storage account.
-
         System.out.println("Creating new vault");
-        Vault vault = azure.vaults().define(VAULT_NAME)
+        vault = azureResourceManager
+                .vaults().define(VAULT_NAME)
                 .withRegion(VAULT_REGION)
                 .withExistingResourceGroup(RESOURCE_GROUP)
                 .defineAccessPolicy()
-                    .forObjectId(USER_OID)
-                    .allowSecretAllPermissions()
-                    .allowStorageAllPermissions()
-                    .attach()
-                .withDeploymentDisabled()
+                .forObjectId(USER_OID)
+                .allowSecretAllPermissions()
+                .allowStorageAllPermissions()
+                .allowCertificateAllPermissions()
+                .allowKeyAllPermissions()
+                .attach()
+                .defineAccessPolicy()
+                .forObjectId(storageAccount.innerModel().identity().principalId())
+                .allowKeyPermissions(KeyPermissions.UNWRAP_KEY,KeyPermissions.WRAP_KEY,KeyPermissions.GET)

```

Comment:
I'm not that familiar with the management plane Key Vault library, do we need to allow these specific permissions if we have allowed all of them above in line 81?

Reply:
These specific permissions are set for establishing an encrypted association between Storage Account and Key Vault. During the processing of `update.withEncryptionKeyFromKeyVault` in line105, the `StorageAccount ` must have permission for access `KayVault`.

Reply:
It has been fixed in new version.